lowcar/devices/KoalaBear/KoalaBear.cpp

```diff
@@ -5,358 +5,346 @@
-
-	this->deadbands = (float *) calloc(2, sizeof(float));
-	this->deadbands[MTRA] = 0.05;
-	this->deadbands[MTRB] = 0.05;
+//******************************* ENCODER TICK INTERRPUT HANDLERS ***************************//
 
-	//this->pid = new PID(0.0, 1.0, 0.0, 0.0, (float) millis());
-	this->pid_enabled = (uint8_t *) calloc(2, sizeof(uint8_t));
-
-	this->prev_led_time = millis();
-	this->curr_led_mtr = MTRA;
-
-	this->led = new LEDKoala();
-	// TODO: Initialize encoder and PID
+// TODO: write a descriptive comment here once we know this works
+static void handle_enc_a_tick() {
+    if (digitalRead(AENC2)) {
+        enc_a++;
+    } else {
+        enc_a--;
+    }
 }
 
-size_t KoalaBear::device_read (uint8_t param, uint8_t *data_buf) {
-	float *float_buf = (float *) data_buf;
-	bool *bool_buf = (bool *) data_buf;
-
-	switch (param) {
-
-		case DUTY_CYCLE_A:
-			float_buf[0] = this->desired_speeds[MTRA];
-			return sizeof(float);
-
-		case DEADBAND_A:
-			float_buf[0] = this->deadbands[MTRA];
-			return sizeof(float);
-
-		case CURRENT_A:
-			// TODO
-			return sizeof(float);
-
-		case PID_ENABLED_A:
-			float_buf[0] = this->pid_enabled[MTRA];
-			return sizeof(uint8_t);
-
-		//TODO: ask PID controller for motor A for KP
-		case PID_KP_A:
-			return sizeof(float);
-
-		//TODO: ask PID controller for motor A for KI
-		case PID_KI_A:
-			return sizeof(float);
-
-		//TODO: ask PID controller for motor A for KD
-		case PID_KD_A:
-			return sizeof(float);
-
-		case ENC_A:
-			//float_buf[0] = this->pid->readPos();
-			return sizeof(float);
-
-		// Params for Motor B
-
-		case DUTY_CYCLE_B:
-			float_buf[0] = this->desired_speeds[MTRB];
-			return sizeof(float);
-
-		case DEADBAND_B:
-			float_buf[0] = this->deadbands[MTRB];
-			return sizeof(float);
-
-		case CURRENT_B:
-			// TODO
-			return sizeof(float);
-
-		case PID_ENABLED_B:
-			data_buf[0] = this->pid_enabled[MTRB];
-			return sizeof(uint8_t);
-
-		//TODO: ask PID controller for motor B for KP
-		case PID_KP_B:
-			return sizeof(float);
+// TODO: write a descriptive comment here once we know this works
+static void handle_enc_b_tick() {
+    if (digitalRead(BENC2)) {
+        enc_b++;
+    } else {
+        enc_b--;
+    }
+}
 
-		//TODO: ask PID controller for motor B for KI
-		case PID_KI_B:
-			return sizeof(float);
+//*********************************** MAIN KOALABEAR CODE ***********************************//
 
-		//TODO: ask PID controller for motor B for KD
-		case PID_KD_B:
-			return sizeof(float);
+KoalaBear::KoalaBear() : Device(DeviceType::KOALA_BEAR, 13) {
+    // initialize motor values    
+    this->target_speed_a = this->target_speed_b = 0.0;
+    this->deadband_a = this->deadband_b = 0.05;
+    
+    // initialize encoders
+    enc_a = enc_b = 0.0;
+    attachInterrupt(digitalPinToInterrupt(AENC1), handle_enc_a_tick, RISING); // set interrupt service routines for encoder pins
+    attachInterrupt(digitalPinToInterrupt(BENC1), handle_enc_b_tick, RISING);
```

Comment:
I think we'll eventually need 4 total interrupts

Reply:
I talked to Sherwin about this, because I was confused as well (he told me how to program this). We won't need 4 total interrupts, because we each encoder is connected to two pins; one is the pin to interrupt on and the other is the pin to read which direction the motor is going. I can explain on Saturday, it's hard to explain in writing.

Reply:
Ok I understand